### PR TITLE
Chore: provider unit tests (w/o ginkgo)

### LINKED ISSUE #68: Project Description isn't written on Create
1. Create new project, with description in TF, like - 

```
resource "env0_project" "my_project" {
  name        = "Provider Project"
  description = "Project Created By Provider"
}
```

2. apply
3. The description isn't applied.

If you update, the description will be there, but not on create.

### PR DESCRIPTION
**Pull request by yaronya**
### Issue & Steps to Reproduce / Feature Request
Add unit tests to our provider package w/o creating remote resources.

### Solution
1. Add unit test to the provider package (i.e `env0`) by mocking the API client using `gomock`.
2. Change the `ApiClient` struct usage to the new `ApiClientInterface` (following the guideline of using interfaces and not structs) - this allows us to mock the API client easily w/o failing on any interface conversions done to the `meta` object of provider.
3. Create a custom `TestReporter` implementing the `testing.T` interface while overriding `Fatalf` func (used by `gomock`) to exit the process instead of just terminating the goroutine (otherwise tests hangs forever when there's a failure on any mock)
4. While testing I found out that `ProjectCreate` did not set the `description` attribute correctly - fixed it. (Fixes #68)

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -15,4 +15,4 @@ jobs:
       - name: Checkout code
         uses: actions/checkout@v2
       - name: Go Test
-        run: go test ./...
\ No newline at end of file
+        run: go test -v ./...
```
Comment (razbensimon) [inline .github/workflows/ci.yml:18]: please add `\n` in EOF

File: client/project.go [hunk 2]
```diff
@@ -22,14 +22,14 @@ func (self *ApiClient) Project(id string) (Project, error) {
 	return result, nil
 }
 
-func (self *ApiClient) ProjectCreate(name string) (Project, error) {
+func (self *ApiClient) ProjectCreate(name string, description string) (Project, error) {
```
Comment (razbensimon) [inline client/project.go:25]: i think we should pass an options object that contains the fields, as we did in other places

File: client/project_test.go [hunk 3]
```diff
@@ -72,7 +74,7 @@ var _ = Describe("Project", func() {
 			mockedResponse.Description = payload.Description
 
 			httpCall = mockHttpClient.EXPECT().
-				Put("/projects/" + mockProject.Id, payload, gomock.Any()).
+				Put("/projects/"+mockProject.Id, payload, gomock.Any()).
```
Comment (razbensimon) [inline client/project_test.go:77]: why no spaces?
you did it on other files too

File: env0/provider_test.go [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+	ctrl = gomock.NewController(&tr)
+
+	apiClientMock = client.NewMockApiClientInterface(ctrl)
+	mockFunc(apiClientMock)
+
+	c.ProviderFactories = testUnitProviders
+	resource.UnitTest(&tr, c)
+
+	ctrl.Finish()
+}
```
Comment (razbensimon) [inline env0/provider_test.go:39]:
```suggestion
func runUnitTest(test *testing.T, testCase resource.TestCase, mockFunc func(mockFunc *client.MockApiClientInterface)) {
        testReporter := utils.TestReporter{T: test}

	ctrl = gomock.NewController(&testReporter)

	apiClientMock = client.NewMockApiClientInterface(ctrl)
	mockFunc(apiClientMock)

	testCase.ProviderFactories = testUnitProviders
	resource.UnitTest(&testReporter, testCase)

	ctrl.Finish()
}
```

What are you stingy about? 💸

File: env0/resource_project_test.go [hunk 5]
```diff
@@ -0,0 +1,66 @@
+package env0
+
+import (
+	"fmt"
+	"github.com/env0/terraform-provider-env0/client"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+	"testing"
+)
+
+func TestUnitProjectResource(t *testing.T) {
```
Comment (razbensimon) [inline env0/resource_project_test.go:10]:
```suggestion
func TestUnitProjectResource(test *testing.T) {
```

Comment (razbensimon) [inline env0/resource_project_test.go:10]:
```suggestion
func TestProjectResource(t *testing.T) {
```

File: env0/provider_test.go [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+
+var testUnitProviders = map[string]func() (*schema.Provider, error){
+	"env0": func() (*schema.Provider, error) {
+		provider := Provider()
+		provider.ConfigureFunc = func(d *schema.ResourceData) (interface{}, error) {
+			return apiClientMock, nil
+		}
+		return provider, nil
+	},
+}
```
Comment (razbensimon) [inline env0/provider_test.go:25]: Can you explain more about this logic? and what is the goal?

**Review by razbensimon — COMMENTED** [review]
Looks Good! 🎸 
most of my comments are renames

File: utils/test_reporter.go [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+See issue in: https://github.com/golang/mock/issues/145
+*/
+
+type TestReporter struct {
+	T *testing.T
+}
+
+func (r *TestReporter) Error(args ...interface{}) {
+	r.T.Error(args...)
+}
```
Comment (razbensimon) [inline utils/test_reporter.go:20]: Please change those too

```suggestion
type TestReporter struct {
	test *testing.T
}

func (reporter *TestReporter) Error(args ...interface{}) {
	reporter.T.Error(args...)
}
```

File: client/project.go (hunk 2 above)
Comment (yaronya) [reply client/project.go:25]: Feels a little out of scope for this PR.
I'll open an issue for that as it's also relevant for the configuration variable parameters.

File: client/project_test.go (hunk 3 above)
Comment (yaronya) [reply client/project_test.go:77]: That's GoLand doing it 🤷 
Guess it's per go specification

File: env0/resource_project_test.go (hunk 5 above)
Comment (yaronya) [reply env0/resource_project_test.go:10]: That's go specification, the signature is always `t *testingT`

Comment (yaronya) [reply env0/resource_project_test.go:10]: That's go specification, the signature is always `t *testingT`

File: utils/test_reporter.go (hunk 7 above)
Comment (yaronya) [reply utils/test_reporter.go:20]: This is a common pattern in go. Usually "`this`" receives a single letter or some generic name.

File: env0/provider_test.go (hunk 6 above)
Comment (yaronya) [reply env0/provider_test.go:25]: terraform's `resource.TestCase` expects the get a `providerFactory`. Here what I do is I create the provider factory but I override its `configureFunc` (that is in charge of setting the `meta` object which the API client sits on) to return the mocked API client.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (yaronya) [reply .github/workflows/ci.yml]: 91bdf62

File: env0/provider_test.go (hunk 4 above)
Comment (yaronya) [reply env0/provider_test.go]: 435e081

File: env0/resource_project_test.go (hunk 5 above)
Comment (razbensimon) [reply env0/resource_project_test.go:10]: I just asked to remove the "Unit" word from the name

**Review by razbensimon — APPROVED** [review]

**Comment by yaronya** [comment]
[Add a proper sample unit test for API client/Provider](https://trello.com/c/psM3ecDW/3297-add-a-proper-sample-unit-test-for-api-client-provider)

**Comment by yaronya** [comment]
[UT for env0/resource_project.go](https://trello.com/c/4xBJrNjM/3365-ut-for-env0-resourceprojectgo)

**Pull request merged**